arch/risc-v/src/common/riscv_swint.c

```diff
@@ -497,8 +515,16 @@ int riscv_swint(int irq, void *context, void *arg)
           if (index == 0 && rtcb->xcp.kstack != NULL)
             {
               rtcb->xcp.ustkptr = (uintptr_t *)regs[REG_SP];
-              regs[REG_SP]      = (uintptr_t)rtcb->xcp.kstack +
+              if (rtcb->xcp.kstkptr == NULL)
```

Comment:
```suggestion
              if (rtcb->xcp.kstkptr != NULL)
```

Reply:
or remove the check since it has been done at line 515

Reply:
or change to DEBUGASSERT like line 453

Reply:
The original check (== NULL) is actually correct. However, the original implementation is very difficult to follow and understand. I can't know how it was supposed to work, only that it was broken.

Without this check, i.e. unconditionally setting 

`regs[REG_SP]  = (uintptr_t)rtcb->xcp.kstack + ARCH_KERNEL_STACKSIZE;`

Destroys the entire kernel stack frame, because it resets the stack pointer to stack top. If rtcb->xcp.kstkptr == NULL we know that no one is using the kernel stack atm and we can set the stack to stack top. Otherwise we should not touch regs[REG_SP].

Reply:
I spent a bit of time on this and I think the new patch should now do what this was originally intended to do:
- If rtcb->xcp.kstkptr is set, it means someone is already using the kernel stack and regs[SP] should be set to this
- If rtcb->xcp.kstkptr is not set, it means no one is using it, and regs[SP] should be set to the top of the kernel stack

Tested and verified the signal dispatch, and calling any system call from the signal handler still works

Reply:
The change looks good could you apply the similar change to arm/xtensa?

Reply:
Done for ARM.

I could not find kernel stack related code from xtensa or any other platform

Reply:
Thanks @pussuw. I will merge your patch once it pass ci.